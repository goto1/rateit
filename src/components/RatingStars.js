import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { Icon } from "./f7";

const Wrapper = styled.div`width: 95px;`;

const IconStyled = styled(Icon)`
  font-size: 19px;
  color: #FEBA72;
  width: 19px;
  padding-bottom: 2px;
`;

const getCombinationOfStars = rating => {
  const TOTAL_NUMBER_OF_STARS = 5;
  const ratingStringified = rating.toString();
  const decimalPointIndex = ratingStringified.indexOf(".");

  const halfStar = decimalPointIndex > -1 ? 1 : 0;
  const wholeStars = halfStar
    ? +ratingStringified.slice(0, decimalPointIndex)
    : +ratingStringified;
  const emptyStars = TOTAL_NUMBER_OF_STARS - halfStar - wholeStars;

  return { halfStar, wholeStars, emptyStars };
};

const RatingStars = ({ rating }) => {
  const { halfStar, wholeStars, emptyStars } = getCombinationOfStars(rating);
  return (
    <Wrapper>
      {Array.apply(null, Array(wholeStars)).map((item, idx) =>
        <IconStyled key={idx} material="star" />
      )}

      {halfStar === 1 && <IconStyled material="star_half" />}

      {Array.apply(null, Array(emptyStars)).map((item, idx) =>
        <IconStyled key={idx} material="star_border" />
      )}
    </Wrapper>
  );
};

RatingStars.propTypes = {
  rating: PropTypes.number.isRequired
};

export default RatingStars;
