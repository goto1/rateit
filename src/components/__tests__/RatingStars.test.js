import React from "react";
import RatingStars from "../RatingStars";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

const getWrapper = rating => shallow(<RatingStars rating={rating} />);

describe("RatingStars component", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<RatingStars rating={4} />);
    expect(tree).toMatchSnapshot();
  });

  it("renders the correct amount of stars", () => {
    const wrapper = getWrapper(4);
    expect(wrapper.children()).toHaveLength(5);
  });

  it("renders the correct amount of empty stars", () => {
    const wrapper = getWrapper(2.5);
    const emptyStars = wrapper
      .children()
      .filterWhere(star => star.render().text() === "star_border");

    expect(emptyStars).toHaveLength(2);
  });

  it("renders the correct amount of half stars", () => {
    const wrapper = getWrapper(3.5);
    const halfStars = wrapper
      .children()
      .filterWhere(star => star.render().text() === "star_half");

    expect(halfStars).toHaveLength(1);
  });

  it("renders the correct amount of full stars", () => {
    const wrapper = getWrapper(3.5);
    const fullStars = wrapper
      .children()
      .filterWhere(star => star.render().text() === "star");

    expect(fullStars).toHaveLength(3);
  });
});
