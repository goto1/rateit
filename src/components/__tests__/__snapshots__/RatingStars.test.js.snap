// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingStars component should render correctly 1`] = `
.gDYObI {
  width: 95px;
}

.bHnJJA {
  font-size: 19px;
  color: #FEBA72;
  width: 19px;
  padding-bottom: 2px;
}

<div
  className="sc-iwsKbI gDYObI"
>
  <i
    className="material-icons icon sc-gZMcBi bHnJJA"
  >
    star
  </i>
  <i
    className="material-icons icon sc-gZMcBi bHnJJA"
  >
    star
  </i>
  <i
    className="material-icons icon sc-gZMcBi bHnJJA"
  >
    star
  </i>
  <i
    className="material-icons icon sc-gZMcBi bHnJJA"
  >
    star
  </i>
  <i
    className="material-icons icon sc-gZMcBi bHnJJA"
  >
    star_border
  </i>
</div>
`;
